### PR TITLE
State implementation

### PR DESCRIPTION
**Pull request by J2thatsme**
Added states for GWF, DWS, R718N3, SenlabM, Multisense & Presence mini.
The states are used to calculate usages between two different samples.

### CODE REVIEW

**Review by mariusgiger — COMMENTED** [review]
there are quite a few breaking changes in here - can we solve this without any breaking changes?

**Comment by J2thatsme** [comment]
Make non breaking again. Name new datakeys to ex. relativeCount && remove absoluteCount for the Moment.

File: types/yanzi/presenceMini/uplink.js [hunk 1]
```diff
@@ -14,6 +23,23 @@ function consume(event) {
...
+        event.state.lastMotion,
+        sample.motion,
+      );
+      event.state.lastMotion = sample.motion;
+
+      if (sample.relativeMotion > 1) {
+        sample.occupancy = 2;
+        sample.occupied = true;
+      } else {
+        sample.occupancy = 0;
```
Comment (mariusgiger) [inline types/yanzi/presenceMini/uplink.js:38]: only one motion count in the interval means pending/leaving? possibly add a comment about this assumption

File: types/yanzi/presenceMini/uplink.js [hunk 2]
```diff
@@ -1,3 +1,12 @@
+function incrementValue(lastPulse, pulse) {
+  // Init state && Check for the case the counter reseted
+  if (lastPulse === undefined || lastPulse > pulse) {
+    lastPulse = pulse;
+  }
+  // Calculate increment
+  return pulse - lastPulse;
```
Comment (mariusgiger) [inline types/yanzi/presenceMini/uplink.js:7]: make sure that this is always positive (for all instances of this)

File: types/browan/TBDW100/uplink.js [hunk 3]
```diff
@@ -21,10 +30,13 @@ function consume(event) {
 
   data.temperature = Bits.bitsToUnsigned(bits.substr(17, 7));
   data.temperature -= 32;
-
   data.time = Hex.hexLittleEndianToBigEndian(payload.substr(6, 4), false);
+
   data.count = Hex.hexLittleEndianToBigEndian(payload.substr(10, 6), false);
+  data.relativeCount = incrementValue(event.state.lastCount, data.count);
+  event.state.lastCount = data.count;
```
Comment (mariusgiger) [inline types/browan/TBDW100/uplink.js:37]: init the state?

File: types/elsys/ems-door/reed.schema.json [hunk 4]
```diff
@@ -15,7 +15,26 @@
       "title": "Pulse count absolute",
       "description": "Pulse count absolute",
       "type": "number"
+    },
+    "relativePulse1": {
+      "title": "Increment pulse 1 count",
+      "description": "Increment pulse 1 count",
+      "type": "number"
+    },
+    "pulseAbs2": {
```
Comment (mariusgiger) [inline types/elsys/ems-door/reed.schema.json:24]: align the naming, either pulseRel1 or absolutePulse2

File: types/yanzi/presenceMini/uplink.js [hunk 5]
```diff
@@ -1,3 +1,12 @@
+function incrementValue(lastPulse, pulse) {
+  // Init state && Check for the case the counter reseted
+  if (lastPulse === undefined || lastPulse > pulse) {
```
Comment (mariusgiger) [inline types/yanzi/presenceMini/uplink.js:3]: also handle pulse being undefined

**Review by mariusgiger — COMMENTED** [review]
looks good, a few more comments

File: types/yanzi/presenceMini/uplink.js (hunk 2 above)
Comment (J2thatsme) [reply types/yanzi/presenceMini/uplink.js:7]: if (lastPulse === undefined || lastPulse > pulse) {

Should always catch sensor resets, which then equal 0

File: types/browan/TBDW100/uplink.js (hunk 3 above)
Comment (J2thatsme) [reply types/browan/TBDW100/uplink.js:37]: is handled with line 37

File: types/elsys/ems-door/reed.schema.json (hunk 4 above)
Comment (J2thatsme) [reply types/elsys/ems-door/reed.schema.json:24]: This would be breaking with current datakeys

File: types/yanzi/presenceMini/uplink.js (hunk 5 above)
Comment (J2thatsme) [reply types/yanzi/presenceMini/uplink.js:3]: This gets handled outside the function.

**Review by mariusgiger — APPROVED** [review]

**Pull request merged**